### DISCUSSION TITLE
Depends doesn't unwrap decorators to know if the callable is a function or a generator

### DISCUSSION THREAD

**Discussion by gabrieldemarmiesse** [fastapi/fastapi#9885] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import TypeVar
from collections.abc import Iterator
from fastapi import Depends, FastAPI
from fastapi.testclient import TestClient
from beartype import beartype

Self = TypeVar("Self")


class SomeClass:
    @classmethod
    @beartype    # remove this decorator to make it work
    def my_iterator(cls: type[Self]) -> Iterator[Self]:
        yield cls()


app = FastAPI()


@app.get("/items/", response_model=None)
async def read_items(my_value: SomeClass = Depends(SomeClass.my_iterator)) -> int:
    assert isinstance(my_value, SomeClass)
    return 1


test_client = TestClient(app)

test_client.get("/items")
```

### Description

Run this script with the latest version of the fastapi and beartype packages. Here are the versions I have locally:
```
fastapi==0.100.0
pydantic==1.10.11
beartype==0.14.1
Python 3.10.10
```

If you remove the beartype decorator, you will notice that everything works fine. It seems fastapi can't find out if the original function is a plain function or a generator.

I raised an issue in the beartype repo and the maintainer @leycec believe it's very likely fastapi that can't inspect the callable given to `Depends()`correctly. See the original issue here: https://github.com/beartype/beartype/issues/250

### Operating System

Linux

### Operating System Details

ubuntu 18

### FastAPI Version

fastapi==0.100.0

### Python Version

Python 3.10.10

**Comment by leycec** [comment]
Thanks so much for fielding this on behalf of @beartype, @gabrieldemarmiesse. As always, you amaze and stun in equal measure.

To corroborate that excellent synopsis above, it seems likely (but *not* certain) that FastAPI is failing to unwrap wrappers created by the standard `@functools.wraps` decorator in certain contexts. Since decorators like @beartype leverage `@functools.wraps` to generate the wrappers they create and return, this makes at least a subset of FastAPI incompatible with **decorator chaining** (i.e., composing two or more decorators together).

Thankfully, the solution is *super*-trivial. Well... assuming this is FastAPI's fault, anyway. This might very well be @beartype's fault, but I suspect it's probably not. Anyway, here's a [trivial well-tested utility function in the @beartype codebase that we personally use to unwrap wrappers](https://github.com/beartype/beartype/blob/2b8c6823742c29f76ead1222ad08983082792b02/beartype/_util/func/utilfuncwrap.py#L68). Catch!

```python
from collections.abc import Callable
from typing import Any

def unwrap_func_all(func: Any) -> Callable:
    '''
    Lowest-level **wrappee** (i.e., callable wrapped by the passed wrapper
    callable) of the passed higher-level **wrapper** (i.e., callable wrapping
    the wrappee callable to be returned) if the passed callable is a wrapper
    *or* that callable as is otherwise (i.e., if that callable is *not* a
    wrapper).

    Specifically, this getter iteratively undoes the work performed by:

    * One or more consecutive uses of the :func:`functools.wrap` decorator on
      the wrappee callable to be returned.
    * One or more consecutive calls to the :func:`functools.update_wrapper`
      function on the wrappee callable to be returned.

    Parameters
    ----------
    func : Callable
        Wrapper callable to be unwrapped.

    Returns
    ----------
    Callable
        Either:

        * If the passed callable is a wrapper, the lowest-level wrappee
          callable wrapped by that wrapper.
        * Else, the passed callable as is.
    '''

    # While this callable still wraps another callable, unwrap one layer of
    # wrapping by reducing this wrapper to its next wrappee.
    while hasattr(func, '__wrapped__'):
        func = func.__wrapped__  # type: ignore[attr-defined]

    # Return this wrappee, which is now guaranteed to *NOT* be a wrapper.
    return func
```

As everyone just noticed, @beartype errs on the side of volubility. It's a character flaw. :smiling_face_with_tear:

**Comment by Kludex** [comment]
Can you show me an example which is supposed to work without `beartype` i.e. only with FastAPI? - The most simple example, please.

Also, which are the cases it fails? The title says "function or a generator", which I guess is not every function...

**Reply by gabrieldemarmiesse** [reply]
Sure, here is a minimal example without beartype. The beartype decorator could have been replaced by any simple decorator:

Same as before, if the decorator is removed, then everything works correctly.
```python

from typing import TypeVar
from collections.abc import Iterator
from fastapi import Depends, FastAPI
from fastapi.testclient import TestClient
import functools

Self = TypeVar("Self")


def print_arguments_and_output(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        print("Arguments:", args, kwargs)
        result = func(*args, **kwargs)
        print("Output:", result)
        return result
    return wrapper


class SomeClass:
    @classmethod
    @print_arguments_and_output    # remove this decorator to make it work
    def my_iterator(cls: type[Self]) -> Iterator[Self]:
        yield cls()


# we check that my_iterator is still an iterator (thus proving the decorator didn't change the behavior)
assert isinstance(next(SomeClass.my_iterator()), SomeClass)


app = FastAPI()


@app.get("/items/", response_model=None)
async def read_items(my_value: SomeClass = Depends(SomeClass.my_iterator)) -> int:
    assert isinstance(my_value, SomeClass)
    return 1


test_client = TestClient(app)

test_client.get("/items")
```

**Reply by leycec** [reply]
*...lol.* You are super awesome, @gabrieldemarmiesse. I was just about to suggest extricating @beartype from the situation by replacing @beartype with the standard decorator pattern, because @beartype was only unnecessarily muddying the waters here. Then you telepathically read my mind. Kinda creepy, actually.

Since this clearly has *nothing* to do with @beartype, I'll kindly show myself out the back door and get back to banging on our next stable release. Best of luck with yours, @Kludex and crew!

**Reply by Kludex** [reply]
Also, which are the cases it fails? The title says "function or a generator", which I guess is not every function...

**Reply by gabrieldemarmiesse** [reply]
@Kludex I think you misunderstood the title. Fastapi doesn't know if the underlying callable is a function or a generator. You can refer to the latest code snippet for a failing test case

**Comment by shughes-uk** [comment]
I think this is related the issue i'm having where I want to wrap some of my dependencies in a decorator, which is causing FastAPI to run them in a sync thread, instead of the desired async context.

```
def testwrap(func):
    @functools.wraps(func)
    async def wrapper(*args, **kwargs):
        async for item in func(*args, **kwargs):
            yield item

    return wrapper


@testwrap
async def should_run_async() -> AsyncGenerator[None, None]:
    # Runs async but in different async contexts, causing issues with contextvars
    yield None
```

My general use case is wanting to add a quick tracing decorator (from datadog's `ddtrace` library) so I can track things.

**Reply by YuriiMotov** [reply]
Could you provide an MRE?
What FastAPI version are you using?

**Reply by samantha-earthmover** [reply]
I suspect it was actually fixed by https://github.com/fastapi/fastapi/pull/14448 , i'll take it for a spin
